core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -174,80 +187,96 @@ public void run() {
      * Periodically invoked to keep track of the load.
      * Launches additional nodes if necessary.
      */
-    private synchronized void update() {
-        Jenkins jenkins = Jenkins.getInstance();
-        lastSuggestedReview = System.currentTimeMillis();
+    private void update() {
+        provisioningLock.lock();
+        try {
+            lastSuggestedReview = System.currentTimeMillis();
+            Queue.withLock(new Runnable() {
+                @Override
+                public void run() {
+                    Jenkins jenkins = Jenkins.getInstance();
+                    // clean up the cancelled launch activity, then count the # of executors that we are about to
+                    // bring up.
+
+                    int plannedCapacitySnapshot = 0;
+                    List<PlannedNode> completedLaunches = new ArrayList<PlannedNode>();
+
+                    for (Iterator<PlannedNode> itr = pendingLaunches.iterator(); itr.hasNext(); ) {
+                        PlannedNode f = itr.next();
+                        if (f.future.isDone()) {
+                            completedLaunches.add(f);
+                            itr.remove();
+                        } else {
+                            plannedCapacitySnapshot += f.numExecutors;
+                        }
+                    }
 
-        // clean up the cancelled launch activity, then count the # of executors that we are about to bring up.
-        int plannedCapacitySnapshot = 0;
-        List<PlannedNode> completedLaunches = new ArrayList<PlannedNode>();
+                    for (PlannedNode f : completedLaunches) {
+                        try {
+                            Node node = f.future.get();
+                            for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                cl.onComplete(f, node);
+                            }
 
-        synchronized (pendingLaunches) {
-            for (Iterator<PlannedNode> itr = pendingLaunches.iterator(); itr.hasNext(); ) {
-                PlannedNode f = itr.next();
-                if (f.future.isDone()) {
-                    completedLaunches.add(f);
-                    itr.remove();
-                } else {
-                    plannedCapacitySnapshot += f.numExecutors;
-                }
-            }
-        }
+                            jenkins.addNode(node);
+                            LOGGER.log(Level.INFO,
+                                    "{0} provisioning successfully completed. We have now {1,number,integer} computer"
+                                            + "(s)",
```

Comment:
Awkward formatting, which makes even [this link](https://github.com/jenkinsci/jenkins/pull/1596/files?w=1) show excessive diff hunks.